### PR TITLE
Add the ability to generate Launchable test reports

### PR DESCRIPTION
**Pull request by ono-max**
Ruby core team has granted me the permission to integrate [Launchable](https://www.launchableinc.com/) into ruby/ruby. With the incorporation of Launchable, we can enjoy various advantages, including insights into flakey tests, a Github Comments feature, analysis for flakey tests, a rich UI for viewing tests stderr and stdout in failed tests, and more.

To initiate this process, I have implemented the generation of test reports in Launchable JSON format within this PR.

### CODE REVIEW

File: tool/lib/test/unit.rb [hunk 1]
```diff
@@ -882,6 +911,61 @@ def setup_options(opts, options)
...
+          @file.write(value)
+          write_comma
+        end
+        def close
+          close_array
+          @file.write('}')
+        end
+        private
+        def write_new_line
+          @file.write('\n')
```
Comment (nobu) [inline tool/lib/test/unit.rb:957]:
```suggestion
          @file.write("\n")
```

File: tool/lib/test/unit.rb [hunk 2]
```diff
@@ -882,6 +911,61 @@ def setup_options(opts, options)
...
+      end
+      class JsonStreamWriter # :nodoc: all
+        def initialize(path)
+          @file = File.open(path, 'w')
+          @file.write('{')
+          write_new_line
+        end
+        def batch_write
+          @file.write("    ")
+          @file.write("{")
```
Comment (nobu) [inline tool/lib/test/unit.rb:928]: Why separate calls?

File: tool/lib/test/unit.rb [hunk 3]
```diff
@@ -882,6 +911,61 @@ def setup_options(opts, options)
...
+          write_new_line
+          yield
+          @file.write("}")
+          write_comma
+          write_new_line
+        end
+        def write_array(key)
+          @file.write("  ")
+          @file.write(key)
```
Comment (nobu) [inline tool/lib/test/unit.rb:937]: `IO#write` takes multiple arguments.
```suggestion
          @file.write("  ", key)
```

File: tool/lib/test/unit.rb [hunk 4]
```diff
@@ -882,6 +911,61 @@ def setup_options(opts, options)
...
+          @file.write('{')
+          write_new_line
+        end
+        def batch_write
+          @file.write("    ")
+          @file.write("{")
+          write_new_line
+          yield
+          @file.write("}")
+          write_comma
```
Comment (nobu) [inline tool/lib/test/unit.rb:932]: JSON prohibits a trailing comma, no?

Comment (ono-max) [reply tool/lib/test/unit.rb:932]: Oh, JSON prohibits a trailing comma as you mentioned. thanks

File: tool/lib/test/unit.rb [hunk 5]
```diff
@@ -882,6 +911,94 @@ def setup_options(opts, options)
         opts.on '--most-asserted=N', Integer, 'Show most asserted N tests' do |n|
           options[:most_asserted] = n
         end
+        opts.on '--launchable-test-reports=PATH', String, 'Report test results in Launchable JSON format' do |path|
+          require 'json'
```
Comment (hsbt) [inline tool/lib/test/unit.rb:915]: Is it necessary now? This PR seems to use only `JsonStreamWriter`.

Comment (ono-max) [reply tool/lib/test/unit.rb:918]: I use JSON module at https://github.com/ruby/ruby/pull/9757/files#diff-d58358cd56895151702a7ae7aae9bc0891f2e692a147a46f388e3abb784e354bR985.

Comment (hsbt) [reply tool/lib/test/unit.rb:918]: Ah, I see.

File: tool/lib/test/unit.rb [hunk 6]
```diff
@@ -882,6 +914,94 @@ def setup_options(opts, options)
...
+      class JsonStreamWriter
+        def initialize(path)
+          @file = File.open(path, "w")
+          @file.write("{")
+          @indent_level = 0
+          @is_first_key_val = true
+          @is_first_obj = true
+          write_new_line
+        end
+        def write_object
```
[the remark this answers is not in the document]
Comment (ono-max) [reply tool/lib/test/unit.rb:937]: [7d81b1c](https://github.com/ruby/ruby/pull/9757/commits/7d81b1c9dc956ad4959a0ea4e8bb87902ffd1745)

File: tool/lib/test/unit.rb [hunk 7]
```diff
@@ -854,6 +854,38 @@ def record(suite, method, assertions, time, error)
...
+                when Test::Unit::AssertionFailedError
+                  status = 'TEST_FAILED'
+                  "Failure:\n#{klass}##{meth} [#{location e}]:\n#{e.message}\n"
+                when Timeout::Error
+                  status = 'TEST_FAILED'
+                  "Timeout:\n#{klass}##{meth}\n"
+                else
+                  status = 'TEST_FAILED'
+                  bt = Test::filter_backtrace(e.backtrace).join "\n    "
+                  "Error:\n#{klass}##{meth}:\n#{e.class}: #{e.message.b}\n    #{bt}\n"
```
Comment (nobu) [inline tool/lib/test/unit.rb:878]: What do you expect as `e` here?
This variable seems unassigned yet inside `case error` block.
Maybe intended `error`?

Comment (ono-max) [reply tool/lib/test/unit.rb:878]: > Maybe intended error?
Yes that's right.

Comment (ono-max) [reply tool/lib/test/unit.rb:878]: I've addressed the change at https://github.com/ruby/ruby/pull/9777/files#diff-d58358cd56895151702a7ae7aae9bc0891f2e692a147a46f388e3abb784e354bR1378.

Thank you for pointing it out!

**Comment by junaruga** [comment]
> Ruby core team has granted me the permission to integrate [Launchable](https://www.launchableinc.com/) into ruby/ruby. With the incorporation of Launchable, we can enjoy various advantages, including insights into flakey tests, a Github Comments feature, analysis for flakey tests, a rich UI for viewing tests stderr and stdout in failed tests, and more.

Where I can find the discussion log of the "Ruby core team has granted me the permission to integrate Launchable into ruby/ruby"? [I searched issue tickets by Launchable on Ruby Bug Tracker](https://bugs.ruby-lang.org/search?utf8=%E2%9C%93&q=Launchable&all_words=&all_words=1&titles_only=&issues=1&news=1&documents=1&changesets=1&wiki_pages=1&messages=1&projects=1&attachments=0&options=0&commit=Search), and but I couldn't find it. It doesn't mean I have an objection about it. I just want to understand the context more.

**Comment by junaruga** [comment]
I found communication logs about Launchable in Ruby project's internal Slack channel.

It seems that an event to explain Launchable was held in January. Then the ruby/debug and ruby/vscode-rdbg repositories are already starting to use Launchable.

https://chouseisan.com/s?h=4d6c803c3ea1492b9b61de0ea7668255 (Japanese)

**Comment by junaruga** [comment]
I think this PR should be squashed because it seems that the 2nd commit of the PR is a fix for the 1st commit.

**Pull request merged**